optaplanner-core/src/main/java/org/optaplanner/core/impl/io/jaxb/GenericJaxbIO.java

```diff
@@ -0,0 +1,318 @@
+            saxParser = saxParserFactory.newSAXParser();
+        } catch (ParserConfigurationException | SAXException e) {
+            throw new OptaPlannerXmlSerializationException(errMessage, e);
+        }
+        XMLReader xmlReader;
+        try {
+            xmlReader = saxParser.getXMLReader();
+        } catch (SAXException e) {
+            throw new OptaPlannerXmlSerializationException(errMessage, e);
+        }
+
+        XMLFilter namespaceOverridingXmlFilter = new NamespaceOverridingXmlFilter(xmlReader, elementNamespaceOverrides);
+        namespaceOverridingXmlFilter.setParent(xmlReader);
+
+        // Use UnmarshallerHandler as a content handler for the XML filter.
+        Unmarshaller unmarshaller = createUnmarshaller();
+        UnmarshallerHandler unmarshallerHandler = unmarshaller.getUnmarshallerHandler();
+        namespaceOverridingXmlFilter.setContentHandler(unmarshallerHandler);
+
+        InputSource xmlInputSource = new InputSource(reader);
+        try {
+            // Parse the XML to feed its content into the UnmarshallerHandler.
+            namespaceOverridingXmlFilter.parse(xmlInputSource);
+        } catch (IOException | SAXException e) {
+            throw new OptaPlannerXmlSerializationException(errMessage, e);
+        }
+
+        try {
+            return (T) unmarshallerHandler.getResult();
+        } catch (JAXBException e) {
+            throw new OptaPlannerXmlSerializationException(errMessage, e);
+        }
+    }
+
+    @Override
+    public void write(T root, Writer writer) {
+        Objects.requireNonNull(root);
+        Objects.requireNonNull(writer);
+        DOMResult domResult = new DOMResult();
+        try {
+            marshaller.marshal(root, domResult);
+        } catch (JAXBException jaxbException) {
+            String errMessage = String.format(ERR_MSG_WRITE, rootClass.getName());
+            throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
+        }
+
+        formatXml(new DOMSource(domResult.getNode()), null, writer);
+    }
+
+    public void writeWithoutNamespaces(T root, Writer writer) {
+        Objects.requireNonNull(root);
+        Objects.requireNonNull(writer);
+        DOMResult domResult = new DOMResult();
+        final String errMessage = String.format(ERR_MSG_WRITE, rootClass.getName());
+        try {
+            marshaller.marshal(root, domResult);
+        } catch (JAXBException jaxbException) {
+            throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
+        }
+
+        try (InputStream xsltInputStream = getClass().getResourceAsStream("removeNamespaces.xslt")) {
+            formatXml(new DOMSource(domResult.getNode()), new StreamSource(xsltInputStream), writer);
+        } catch (IOException e) {
+            throw new OptaPlannerXmlSerializationException(errMessage, e);
+        }
+    }
+
+    private void formatXml(Source source, Source transformationTemplate, Writer writer) {
```

Comment:
Do we need to bring in an XSLT formatter to format the JAXB output?
Can we not use this instead?

`marshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, Boolean.TRUE);`

Reply:
Unfortunately, the `Marshaller.JAXB_FORMATTED_OUTPUT` does always 4 spaces of indentation. We use the XSLT to make it 2 (or any other number, if needed).

For the "backward compatibility mode", when writing the elements, we remove their namespaces via an XSLT, so we need a transformation anyway.